src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs

```diff
@@ -1129,4 +1131,155 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
             codeStream.Emit(ILOpcode.call, emitter.NewToken(helper));
         }
     }
+
+    class CustomTypeMarshaller : Marshaller
+    {
+        private ILLocalVariable MarshallerLocalVariable = (ILLocalVariable)(-1);
+
+        private ILLocalVariable InitializeMarshallerVariable()
+        {
+            if (MarshallerLocalVariable != (ILLocalVariable)(-1))
+            {
+                return MarshallerLocalVariable;
+            }
+
+            var marshallerType = MarshalAsDescriptor.CustomMarshallerType;
+            if (marshallerType == null)
+            {
+                // TypeLoadException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
```

Comment:
ThrowHelper.ThrowTypeLoadException?

Reply:
There no type here. Best overload is `void ThrowTypeLoadException(string @namespace, string name, string moduleName)` but I have to deduce namespace and module name. Module name I can pass to `MarshalAsDescriptor` from `EcmaSignatureParser` but isn't it overkill?

Reply:
If you change MarshalAsDescriptor to have a string instead of TypeDesc and do the conversion here, you can ask the conversion helper to throw exception instead of returning null. It will throw the right exception

Reply:
Look at this https://github.com/dotnet/runtime/pull/62921/commits/0fce5bae841a0f13632373eb258bd93f66461894 I take liberty there, but that's seems to be reasonable way. Notice: This case is where no `TypeDesc == null` and `typeName == null` as well.

Reply:
Is there anything problematic with my suggestion though ("If you change MarshalAsDescriptor to have a string instead of TypeDesc and do the conversion here, you can ask the conversion helper to throw exception")? I'm not a fan of `TypeLoadException` with no message when we instead could throw "FileLoadException: assembly X could not be found", or "TypeLoadException: Type X in assembly Y could not be found".

Reply:
This is case for `null`. Null string is still `null` and we are would not have type information. Also your suggested method was my initial attempt and this is what we settle on because there no reasonable way to create TypeDesc inside Marshaller class

Reply:
That’s the case for message “Cannot load null type”

Reply:
That field is populated as `customMarshallerType = _ecmaModule.GetTypeByCustomAttributeTypeName(managedMarshallerTypeName, false);` so this will be null for everything that didn't resolve. Not just a null string.

Maybe we can flip this to `true` and then it will never be null.

Reply:
Changes which I introduce was for this case only 
https://github.com/dotnet/runtime/blob/84926be4243143b7830e78cb7727057ed711a3cc/src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs#L387-L393

Reply:
Now I re-read and understand what you are saying.

Reply:
I did not handle just the case when type name point to separate assembly. I have to add additional test case and honestly do not want to touch that unless that's nesecssary. This seems to be fringe implementation feature. Let somebody else report issue 😄 . I would gladly help, or maybe as followup PR with tests.